## Introduction

For schools with CS, do any attributes of the individual student contribute to whether or not they take CS? 

In [1]:
import pandas as pd
import numpy as np

In [41]:
original_ospi = pd.read_csv('2022_school_pt5.csv',  index_col=0)

In [2]:
ospi_data = pd.read_csv('2022_school_pt5.csv',  index_col=0)

In [3]:
ospi_data.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  ...  \
0  46.752189   Adams           28              3         9          0  ...   
1  46.822710   Adams         1281            169       635         60  ...   
2  47.125588   Adams           60             17        30          8  ...   
3  47.125588   Adams          113             18        57          9  ...   
4  46.411019  Asotin          167              0        89          0  ...   

   D_NoDisability  DC_NoDisability  A_9  AC_9  A_10  AC_10  A_11  AC_11  A_12  \
0              26                3    7     0     4      0     6      1    11   
1            1086              140  382    51   328     44   321     29   250   
2              57               16   18     9    15      2    14      1    13   
3             107               14   32     5    35      3    25      3    21   
4             142                0   16     0    51      0    41      0    59   

   AC_12  
0      2  
1     45  
2      5  
3      7  
4      0  

[5 rows x 50 columns]

In [4]:
ospi_data.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  ...  \
0  46.752189   Adams           28              3         9          0  ...   
1  46.822710   Adams         1281            169       635         60  ...   
2  47.125588   Adams           60             17        30          8  ...   
3  47.125588   Adams          113             18        57          9  ...   
4  46.411019  Asotin          167              0        89          0  ...   

   D_NoDisability  DC_NoDisability  A_9  AC_9  A_10  AC_10  A_11  AC_11  A_12  \
0              26                3    7     0     4      0     6      1    11   
1            1086              140  382    51   328     44   321     29   250   
2              57               16   18     9    15      2    14      1    13   
3             107               14   32     5    35      3    25      3    21   
4             142                0   16     0    51      0    41      0    59   

   AC_12  
0      2  
1     45  
2      5  
3      7  
4      0  

[5 rows x 50 columns]

In [5]:
def low_inc_perc(row):
    return row['I_LowIncome'] / row['AllStudents']

ospi_data['PercentLowIncome'] = ospi_data.apply(low_inc_perc, axis=1)

In [6]:
ospi_data.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  ...  \
0  46.752189   Adams           28              3         9          0  ...   
1  46.822710   Adams         1281            169       635         60  ...   
2  47.125588   Adams           60             17        30          8  ...   
3  47.125588   Adams          113             18        57          9  ...   
4  46.411019  Asotin          167              0        89          0  ...   

   DC_NoDisability  A_9  AC_9  A_10  AC_10  A_11  AC_11  A_12  AC_12  \
0                3    7     0     4      0     6      1    11      2   
1              140  382    51   328     44   321     29   250     45   
2               16   18     9    15      2    14      1    13      5   
3               14   32     5    35      3    25      3    21      7   
4                0   16     0    51      0    41      0    59      0   

   PercentLowIncome  
0          0.607143  
1          0.834504  
2          0.800000  
3          0.415929  
4          0.796407  

[5 rows x 51 columns]

In [7]:
pd.set_option('display.max_columns', None)
ospi_data.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Male  \
0  46.752189   Adams           28              3         9          0      19   
1  46.822710   Adams         1281            169       635         60     646   
2  47.125588   Adams           60             17        30          8      29   
3  47.125588   Adams          113             18        57          9      56   
4  46.411019  Asotin          167              0        89          0      78   

   GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  RC_Asian  \
0        3          0           0         0          0        0         0   
1      109          0           0         0          0        5         1   
2        9          1           0         0          0        0         0   
3        9          0           0         2          2        2         0   
4        0          0           0         1          0        0         0   

   R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  RC_NA  \
0        0         0           2            0      0       0     0      0   
1        2         0        1173          151      0       0     0      0   
2        1         0          23           10      0       0     0      0   
3        0         0           9            0      0       0     0      0   
4        2         0          11            0      0       0     0      0   

   R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  L_NoELL  \
0            2             0       24         3      0       0       28   
1            3             0       98        17    402      54      879   
2            4             2       32         5      7       3       53   
3            3             0       97        16      0       0      113   
4            5             0      148         0      0       0      167   

   LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  IC_NoLowIncome  \
0         3           17             1             11               2   
1       115         1069           142            212              27   
2        14           48            14             12               3   
3        18           47             6             66              12   
4         0          133             0             34               0   

   D_Disability  DC_Disability  D_NoDisability  DC_NoDisability  A_9  AC_9  \
0             2              0              26                3    7     0   
1           195             29            1086              140  382    51   
2             3              1              57               16   18     9   
3             6              4             107               14   32     5   
4            25              0             142                0   16     0   

   A_10  AC_10  A_11  AC_11  A_12  AC_12  PercentLowIncome  
0     4      0     6      1    11      2          0.607143  
1   328     44   321     29   250     45          0.834504  
2    15      2    14      1    13      5          0.800000  
3    35      3    25      3    21      7          0.415929  
4    51      0    41      0    59      0          0.796407

For the following, note that BIPOC does not include Asian in this calculation. The following calculation includes all reported racial identities except for Asian and white.

In [8]:
def number_bipoc(row):
    return (row['R_Native'] + row['R_Black'] + row['R_Hisp_Lat'] + row['R_HPI'] + row['R_TwoOrMore'])

ospi_data['R_BIPOC'] = ospi_data.apply(number_bipoc, axis=1)

In [9]:
def bipoc_perc(row):
    return (row['R_BIPOC']) / row['AllStudents']

ospi_data['PercentBIPOC'] = ospi_data.apply(bipoc_perc, axis=1)

In [10]:
ospi_data['DistrictNumSchools'] = ospi_data.groupby('DistrictCode')['SchoolCode'].transform('size')
ospi_data['DistrictSize'] = ospi_data.groupby('DistrictCode')['AllStudents'].transform('sum')
ospi_data['DistrictLowIncome'] = ospi_data.groupby('DistrictCode')['I_LowIncome'].transform('sum')
ospi_data['DistrictPctLowIncome'] = ospi_data['DistrictLowIncome']/ospi_data['DistrictSize']
ospi_data['DistrictBIPOC'] = ospi_data.groupby('DistrictCode')['R_BIPOC'].transform('sum')
ospi_data['DistrictPctBIPOC'] = ospi_data['DistrictBIPOC']/ospi_data['DistrictSize']

## Preprocessing: Hot One Encoding
*Hot one encoding whether or not a school offers CS*  

In [11]:
# reduced_ospi = reduced_ospi.copy()
ospi_data['CS_Offered'] = np.where(ospi_data['C_AllStudents'] == 0, 0, 1)
ospi_data.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Male  \
0  46.752189   Adams           28              3         9          0      19   
1  46.822710   Adams         1281            169       635         60     646   
2  47.125588   Adams           60             17        30          8      29   
3  47.125588   Adams          113             18        57          9      56   
4  46.411019  Asotin          167              0        89          0      78   

   GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  RC_Asian  \
0        3          0           0         0          0        0         0   
1      109          0           0         0          0        5         1   
2        9          1           0         0          0        0         0   
3        9          0           0         2          2        2         0   
4        0          0           0         1          0        0         0   

   R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  RC_NA  \
0        0         0           2            0      0       0     0      0   
1        2         0        1173          151      0       0     0      0   
2        1         0          23           10      0       0     0      0   
3        0         0           9            0      0       0     0      0   
4        2         0          11            0      0       0     0      0   

   R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  L_NoELL  \
0            2             0       24         3      0       0       28   
1            3             0       98        17    402      54      879   
2            4             2       32         5      7       3       53   
3            3             0       97        16      0       0      113   
4            5             0      148         0      0       0      167   

   LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  IC_NoLowIncome  \
0         3           17             1             11               2   
1       115         1069           142            212              27   
2        14           48            14             12               3   
3        18           47             6             66              12   
4         0          133             0             34               0   

   D_Disability  DC_Disability  D_NoDisability  DC_NoDisability  A_9  AC_9  \
0             2              0              26                3    7     0   
1           195             29            1086              140  382    51   
2             3              1              57               16   18     9   
3             6              4             107               14   32     5   
4            25              0             142                0   16     0   

   A_10  AC_10  A_11  AC_11  A_12  AC_12  PercentLowIncome  R_BIPOC  \
0     4      0     6      1    11      2          0.607143        4   
1   328     44   321     29   250     45          0.834504     1178   
2    15      2    14      1    13      5          0.800000       28   
3    35      3    25      3    21      7          0.415929       14   
4    51      0    41      0    59      0          0.796407       19   

   PercentBIPOC  DistrictNumSchools  DistrictSize  DistrictLowIncome  \
0      0.142857                   1            28                 17   
1      0.919594                   3          1464               1235   
2      0.466667                   1            60                 48   
3      0.123894                   1           113    

## Preprocessing: Percentage of District with CS

In [12]:
ospi_data['DistrictCS'] = ospi_data.groupby('DistrictCode')['CS_Offered'].transform('sum')
ospi_data['DistrictPctCS'] = ospi_data['DistrictCS']/ospi_data['DistrictNumSchools']
ospi_data.head(25)

DistrictCode  SchoolCode                               SchoolName  \
0           1109        3075         Washtucna Elementary/High School   
1           1147        3015                      Othello High School   
2           1158        2903               Lind-Ritzville High School   
3           1160        2132                    Ritzville High School   
4           2250        1617           Educational Opportunity Center   
5           2250        2299        Charles Francis Adams High School   
6           2420        2434                        Asotin Jr Sr High   
7           3017        1884                       Legacy High School   
8           3017        2826                    Kennewick High School   
9           3017        3731                     Kamiakin High School   
10          3017        4007  Benton/Franklin Juvenile Justice Center   
11          3017        4118                   Tri-Tech Skills Center   
12          3017        4484                   Southridge High School   
13          3052        2904            Kiona-Benton City High School   
14          3053        2367                   River View High School   
15          3116        2508                      Prosser High School   
16          3400        3511                     Richland High School   
17          3400        3833                      Hanford High School   
18          3400        4295                  Rivers Edge High School   
19          4019        2623                       Manson High School   
20          4127        3317            Entiat Middle and High School   
21          4129        4260                       Chelan High School   
22          4222        3268                     CASHMERE HIGH SCHOOL   
23          4228        3564                      Cascade High School   
24          4246        1612                             Skill Source   

     Longitude   Latitude  County  AllStudents  C_AllStudents  G_Female  \
0  -118.311231  46.752189   Adams           28              3         9   
1  -119.165246  46.822710   Adams         1281            169       635   
2  -118.292516  47.125588   Adams           60             17        30   
3  -118.292516  47.125588   Adams          113             18        57   
4  -117.057521  46.411019  Asotin          167              0        89   
5  -117.044772  46.410724  Asotin          803             29       389   
6  -117.048210  46.339330  Asotin          177              0        88   
7  -119.184062  46.199694  Benton          716              0       370   
8  -119.126476  46.206332  Benton         1911            203       942   
9  -119.187120  46.216464  Benton         2007            139       969   
10 -119.195874  46.224092  Benton           92              0        23   
11 -119.200948  46.219565  Benton          392              7       208   
12 -119.191400  46.179867  Benton         1894             76       907   
13 -119.491177  46.263821  Benton          464              0       229   
14 -119.037182  46.155253  Benton          311              9       149   
15 -119.767241  46.203649  Benton          972              2       493   
16 -119.288273  46.278435  Benton         2283             68      1103   
17 -119.267841  46.325664  Benton         2033            120      1013   
18 -119.277102  46.272490  Benton          287              1       156   
19 -120.158410  47.884860  Chelan          221             61       107   
20 -120.225980  47.664507  Chelan           87              0        42   
21 -120.077750  47.876985  Chelan          439              0       195   
22 -120.479061  47.513150  Chelan          498             23       251   
23 -120.654659  47.602066  Chelan          414             20       188   
24 -120.316551  47.427368  Chelan           18              0        10   

    GC_Female  G_Male  GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  \
0           0      19        3          0           0         0          0  

## Preprocessing: Excluding schools without CS

In [13]:
only_cs = ospi_data[ospi_data['CS_Offered'] == 1].copy(deep=True)
only_cs.head()

DistrictCode  SchoolCode                         SchoolName   Longitude  \
0          1109        3075   Washtucna Elementary/High School -118.311231   
1          1147        3015                Othello High School -119.165246   
2          1158        2903         Lind-Ritzville High School -118.292516   
3          1160        2132              Ritzville High School -118.292516   
5          2250        2299  Charles Francis Adams High School -117.044772   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Male  \
0  46.752189   Adams           28              3         9          0      19   
1  46.822710   Adams         1281            169       635         60     646   
2  47.125588   Adams           60             17        30          8      29   
3  47.125588   Adams          113             18        57          9      56   
5  46.410724  Asotin          803             29       389         10     413   

   GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  RC_Asian  \
0        3          0           0         0          0        0         0   
1      109          0           0         0          0        5         1   
2        9          1           0         0          0        0         0   
3        9          0           0         2          2        2         0   
5       19          1           0        19          0        9         0   

   R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  RC_NA  \
0        0         0           2            0      0       0     0      0   
1        2         0        1173          151      0       0     0      0   
2        1         0          23           10      0       0     0      0   
3        0         0           9            0      0       0     0      0   
5       14         0          74            2      0       0     0      0   

   R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  L_NoELL  \
0            2             0       24         3      0       0       28   
1            3             0       98        17    402      54      879   
2            4             2       32         5      7       3       53   
3            3             0       97        16      0       0      113   
5           56             4      631        23      4       0      799   

   LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  IC_NoLowIncome  \
0         3           17             1             11               2   
1       115         1069           142            212              27   
2        14           48            14             12               3   
3        18           47             6             66              12   
5        29          389            11            414              18   

   D_Disability  DC_Disability  D_NoDisability  DC_NoDisability  A_9  AC_9  \
0             2              0              26                3    7     0   
1           195             29            1086              140  382    51   
2             3              1              57               16   18     9   
3             6              4             107               14   32     5   
5           117              1             686               28  220    13   

   A_10  AC_10  A_11  AC_11  A_12  AC_12  PercentLowIncome  R_BIPOC  \
0     4      0     6      1    11      2          0.607143        4   
1   328     44   321     29   250     45          0.834504     1178   
2    15      2    14      1    13      5          0.800000       28   
3    35      3    25      3    21      7          0.415929       14   
5   218     12   179      3   186      1          0.484433      163   

   PercentBIPOC  DistrictNumSchools  DistrictSize  DistrictLowIncome  \
0      0.142857                   1            28                 17   
1      0.919594                   3          1464               1235   
2      0.466667                   1            60                 48   
3      0.123894                   1           1

## Preprocessing: Generating synthetic student records
*Using probabilities found in the dataset*

In [14]:
synth_school = only_cs.copy(deep=True)

In [15]:
# Calculating probabilities for each school

# Male

def calc_male_prob(row):
    return row['G_Male'] / row['AllStudents']

synth_school['G_Male_Prob'] = synth_school.apply(calc_male_prob, axis=1)


# Native

def calc_native_prob(row):
    return row['R_Native'] / row['AllStudents']

synth_school['R_Native_Prob'] = synth_school.apply(calc_native_prob, axis=1)

# Asian

def calc_asian_prob(row):
    return row['R_Asian'] / row['AllStudents']

synth_school['R_Asian_Prob'] = synth_school.apply(calc_asian_prob, axis=1)

# Black

def calc_black_prob(row):
    return row['R_Black'] / row['AllStudents']

synth_school['R_Black_Prob'] = synth_school.apply(calc_black_prob, axis=1)

# Hisp_Lat

def calc_hisp_lat_prob(row):
    return row['R_Hisp_Lat'] / row['AllStudents']

synth_school['R_Hisp_Lat_Prob'] = synth_school.apply(calc_hisp_lat_prob, axis=1)

# HPI

def calc_hpi_prob(row):
    return row['R_HPI'] / row['AllStudents']

synth_school['R_HPI_Prob'] = synth_school.apply(calc_hpi_prob, axis=1)

# R_NA

def calc_na_prob(row):
    return row['R_NA'] / row['AllStudents']

synth_school['R_NA_Prob'] = synth_school.apply(calc_na_prob, axis=1)

# R_TwoOrMore

def calc_twoormore_prob(row):
    return row['R_TwoOrMore'] / row['AllStudents']

synth_school['R_TwoOrMore_Prob'] = synth_school.apply(calc_twoormore_prob, axis=1)

# White

def calc_white_prob(row):
    return row['R_White'] / row['AllStudents']

synth_school['R_White_Prob'] = synth_school.apply(calc_white_prob, axis=1)

# ELL

def calc_ell_prob(row):
    return row['L_ELL'] / row['AllStudents']

synth_school['L_ELL_Prob'] = synth_school.apply(calc_ell_prob, axis=1)

# No ELL

def calc_noell_prob(row):
    return row['L_NoELL'] / row['AllStudents']

synth_school['L_NoELL_Prob'] = synth_school.apply(calc_noell_prob, axis=1)

# Low Income

def calc_lowincome_prob(row):
    return row['I_LowIncome'] / row['AllStudents']

synth_school['I_LowIncome_Prob'] = synth_school.apply(calc_lowincome_prob, axis=1)

# No Low Income

def calc_nolow_prob(row):
    return row['I_NoLowIncome'] / row['AllStudents']

synth_school['I_NoLowIncome_Prob'] = synth_school.apply(calc_nolow_prob, axis=1)

# Disability

def calc_disability_prob(row):
    return row['D_Disability'] / row['AllStudents']

synth_school['D_Disability_Prob'] = synth_school.apply(calc_disability_prob, axis=1)

# No Disability

def calc_nodisability_prob(row):
    return row['D_NoDisability'] / row['AllStudents']

synth_school['D_NoDisability_Prob'] = synth_school.apply(calc_nodisability_prob, axis=1)

# Grade 9

def calc_a9_prob(row):
    return row['A_9'] / row['AllStudents']

synth_school['A_9_Prob'] = synth_school.apply(calc_a9_prob, axis=1)

# Grade 10

def calc_a10_prob(row):
    return row['A_10'] / row['AllStudents']

synth_school['A_10_Prob'] = synth_school.apply(calc_a10_prob, axis=1)

# Grade 11

def calc_a11_prob(row):
    return row['A_11'] / row['AllStudents']

synth_school['A_11_Prob'] = synth_school.apply(calc_a11_prob, axis=1)

# Grade 12

def calc_a12_prob(row):
    return row['A_12'] / row['AllStudents']

synth_school['A_12_Prob'] = synth_school.apply(calc_a12_prob, axis=1)

In [16]:
# Calculating CS probabilities for each school

def calc_fem_or_x_cs_prob(row):
    if (row['G_Female'] + row['G_GenderX']) == 0:
        return 0
    else:
        return (row['GC_Female'] + row['GC_GenderX']) / (row['G_Female'] + row['G_GenderX'])

synth_school['GC_Female_or_X_Prob'] = synth_school.apply(calc_fem_or_x_cs_prob, axis=1)

# Male

def calc_male_cs_prob(row):
    return row['GC_Male'] / row['G_Male']

synth_school['GC_Male_Prob'] = synth_school.apply(calc_male_cs_prob, axis=1)

# Native

def calc_native_cs_prob(row):
    if row['R_Native'] == 0:
        return 0
    else:
        return row['RC_Native'] / row['R_Native']

synth_school['RC_Native_Prob'] = synth_school.apply(calc_native_cs_prob, axis=1)

# Asian

def calc_asian_cs_prob(row):
    if row['R_Asian'] == 0:
        return 0
    else:
        return row['RC_Asian'] / row['R_Asian']

synth_school['RC_Asian_Prob'] = synth_school.apply(calc_asian_cs_prob, axis=1)

# Black

def calc_black_cs_prob(row):
    if row['R_Black'] == 0:
        return 0
    else:
        return row['RC_Black'] / row['R_Black']

synth_school['RC_Black_Prob'] = synth_school.apply(calc_black_cs_prob, axis=1)

# Hisp_Lat

def calc_hisp_lat_cs_prob(row):
    if row['R_Hisp_Lat'] == 0:
        return 0
    else:
        return row['RC_Hisp_Lat'] / row['R_Hisp_Lat']

synth_school['RC_Hisp_Lat_Prob'] = synth_school.apply(calc_hisp_lat_cs_prob, axis=1)

# HPI

def calc_hpi_cs_prob(row):
    if row['R_HPI'] == 0:
        return 0
    else:
        return row['RC_HPI'] / row['R_HPI']

synth_school['RC_HPI_Prob'] = synth_school.apply(calc_hpi_cs_prob, axis=1)

# RC_NA

def calc_na_cs_prob(row):
    if row['R_NA'] == 0:
        return 0
    else:
        return row['RC_NA'] / row['R_NA']

synth_school['RC_NA_Prob'] = synth_school.apply(calc_na_cs_prob, axis=1)

# RC_TwoOrMore

def calc_twoormore_cs_prob(row):
    if row['R_TwoOrMore'] == 0:
        return 0
    else:
        return row['RC_TwoOrMore'] / row['R_TwoOrMore']

synth_school['RC_TwoOrMore_Prob'] = synth_school.apply(calc_twoormore_cs_prob, axis=1)

# White

def calc_white_cs_prob(row):
    return row['RC_White'] / row['R_White']

synth_school['RC_White_Prob'] = synth_school.apply(calc_white_cs_prob, axis=1)

# ELL

def calc_ell_cs_prob(row):
    if row['L_ELL'] == 0:
        return 0
    else:
        return row['LC_ELL'] / row['L_ELL']

synth_school['LC_ELL_Prob'] = synth_school.apply(calc_ell_cs_prob, axis=1)

# No ELL

def calc_noell_cs_prob(row):
    return row['LC_NoELL'] / row['L_NoELL']

synth_school['LC_NoELL_Prob'] = synth_school.apply(calc_noell_cs_prob, axis=1)

# Low Income

def calc_lowincome_cs_prob(row):
    if row['I_LowIncome'] == 0:
        return 0
    else:
        return row['IC_LowIncome'] / row['I_LowIncome']

synth_school['IC_LowIncome_Prob'] = synth_school.apply(calc_lowincome_cs_prob, axis=1)

# No Low Income

def calc_nolow_cs_prob(row):
    if row['I_NoLowIncome'] == 0:
        return 0
    else:
        return row['IC_NoLowIncome'] / row['I_NoLowIncome']

synth_school['IC_NoLowIncome_Prob'] = synth_school.apply(calc_nolow_cs_prob, axis=1)

# Disability

def calc_disability_cs_prob(row):
    if row['D_Disability'] == 0:
        return 0
    else:
        return row['DC_Disability'] / row['D_Disability']

synth_school['DC_Disability_Prob'] = synth_school.apply(calc_disability_cs_prob, axis=1)

# No Disability

def calc_nodisability_cs_prob(row):
    if row['D_NoDisability'] == 0:
        return 0
    else:
        return row['DC_NoDisability'] / row['D_NoDisability']

synth_school['DC_NoDisability_Prob'] = synth_school.apply(calc_nodisability_cs_prob, axis=1)

# Grade 9

def calc_a9_cs_prob(row):
    if row['A_9'] == 0:
        return 0
    else:
        return row['AC_9'] / row['A_9']

synth_school['AC_9_Prob'] = synth_school.apply(calc_a9_cs_prob, axis=1)

# Grade 10

def calc_a10_cs_prob(row):
    if row['A_10'] == 0:
        return 0
    else:
        return row['AC_10'] / row['A_10']

synth_school['AC_10_Prob'] = synth_school.apply(calc_a10_cs_prob, axis=1)

# Grade 11

def calc_a11_cs_prob(row):
    if row['A_11'] == 0:
        return 0
    else:
        return row['AC_11'] / row['A_11']

synth_school['AC_11_Prob'] = synth_school.apply(calc_a11_cs_prob, axis=1)

# Grade 12

def calc_a12_cs_prob(row):
    if row['A_12'] == 0:
        return 0
    else:
        return row['AC_12'] / row['A_12']

synth_school['AC_12_Prob'] = synth_school.apply(calc_a12_cs_prob, axis=1)


In [17]:
def calc_all_cs_cs_prob(row):
    return row['C_AllStudents'] / row['AllStudents']

synth_school['C_AllStudents_Prob'] = synth_school.apply(calc_all_cs_cs_prob, axis=1)

In [18]:
synth_students = synth_school.loc[synth_school.index.repeat(synth_school.AllStudents)].copy(deep=True)

In [19]:
rng2 = np.random.default_rng(seed=2024)


def student_gender(row):

    g_probs = np.array([row['G_Male_Prob'], 1-row['G_Male_Prob']])
    
    g_probs /= np.sum(g_probs)
    
    return rng2.choice([1, 0], size = 1, p = g_probs)
    
synth_students['male'] = synth_students.apply(student_gender, axis=1)

In [20]:
def student_race(row):
    
    r_probs = np.array([row['R_Native_Prob'], row['R_Asian_Prob'], 
                        row['R_Black_Prob'], row['R_Hisp_Lat_Prob'], 
                        row['R_HPI_Prob'], row['R_NA_Prob'], 
                        row['R_TwoOrMore_Prob'], row['R_White_Prob']], dtype=float)
    r_probs /= np.sum(r_probs)
    
    return rng2.choice(['Native', 'Asian', 'Black', 'Hisp_Lat', 'HPI', 'NA', 'TwoOrMore', 'White'], size = 1, p = r_probs)

synth_students['race'] = synth_students.apply(student_race, axis=1)

In [21]:
def student_ell(row):
    
    e_probs = np.array([row['L_ELL_Prob'], row['L_NoELL_Prob']], dtype=float)
    
    e_probs /= np.sum(e_probs)
    
    return rng2.choice([1, 0], size = 1, p = e_probs)

synth_students['ell'] = synth_students.apply(student_ell, axis=1)

In [22]:
def student_low_income(row):
    
    i_probs = np.array([row['I_LowIncome_Prob'], row['I_NoLowIncome_Prob']], dtype=float)
    
    i_probs /= np.sum(i_probs)
    
    return rng2.choice([1, 0], size = 1, p = i_probs)

synth_students['low_income'] = synth_students.apply(student_low_income, axis=1)

In [23]:
def student_disability(row):
    
    d_probs = np.array([row['D_Disability_Prob'], row['D_NoDisability_Prob']], dtype=float)
    
    d_probs /= np.sum(d_probs)
    
    return rng2.choice([1, 0], size = 1, p = d_probs)

synth_students['disability'] = synth_students.apply(student_disability, axis=1)

In [24]:
def student_grade(row):
    
    a_probs = np.array([row['A_9_Prob'], row['A_10_Prob'], row['A_11_Prob'], row['A_12_Prob']], dtype=float)
    
    a_probs /= np.sum(a_probs)
    
    return rng2.choice(['9','10','11','12'], size = 1, p = a_probs)

synth_students['grade'] = synth_students.apply(student_grade, axis=1)

In [25]:
def cs_gender_prob(row):
    if row['male'] == 1:
        return row['GC_Male_Prob']
    else:
        return row['GC_Female_or_X_Prob']
        
synth_students['gender_cs_prob'] = synth_students.apply(cs_gender_prob, axis=1)

In [26]:
def cs_race_prob(row):
    if row['race'] == 'Asian':
        return row['RC_Native_Prob']
    if row['race'] == 'Black':
         return row['RC_Black_Prob']
    if row['race'] == 'Hisp_Lat':        
        return row['RC_Hisp_Lat_Prob']
    if row['race'] == 'HPI':
        return row['RC_HPI_Prob']
    if row['race'] == 'NA':
        return row['RC_NA_Prob']
    if row['race'] == 'TwoOrMore':
        return row['RC_TwoOrMore_Prob']
    if row['race'] == 'Asian':
        return row['RC_Asian_Prob']
    elif row['race'] == 'White':
        return row['RC_White_Prob']
        
synth_students['race_cs_prob'] = synth_students.apply(cs_race_prob, axis=1)

In [27]:
def cs_ell_prob(row):
    if row['ell'] == 1:
        return row['LC_ELL_Prob']
    
    elif row['ell'] == 0:
        return row['LC_NoELL_Prob']
    
        
synth_students['ell_cs_prob'] = synth_students.apply(cs_ell_prob, axis=1)

In [28]:
def cs_low_income_prob(row):
    if row['low_income'] == 1:
        return row['IC_LowIncome_Prob']
    
    elif row['low_income'] == 0:
        return row['IC_NoLowIncome_Prob']
    
synth_students['low_income_cs_prob'] = synth_students.apply(cs_low_income_prob, axis=1)

In [29]:
def cs_disability_prob(row):
    if row['disability'] == 1:
        return row['DC_Disability_Prob']
    
    elif row['disability'] == 0:
        return row['DC_NoDisability_Prob']
    
synth_students['disability_cs_prob'] = synth_students.apply(cs_disability_prob, axis=1)

In [30]:
def cs_grade_prob(row):
    if row['grade'] == '9':
        return row['AC_9_Prob']
    
    elif row['grade'] == '10':
        return row['AC_10_Prob']
    
    elif row['grade'] == '11':
        return row['AC_11_Prob']
    
    elif row['grade'] == '12':
        return row['AC_12_Prob']
        
synth_students['grade_cs_prob'] = synth_students.apply(cs_grade_prob, axis=1)

In [31]:
synth_students.fillna(0, inplace=True)

In [32]:
def cs_prob(row):
    return (row['gender_cs_prob'] + row['race_cs_prob'] + row['ell_cs_prob'] + row['low_income_cs_prob'] + row['disability_cs_prob'] + row['grade_cs_prob']) / 6

synth_students['CS_Prob'] = synth_students.apply(cs_prob, axis=1)

In [33]:
synth_students[synth_students['CS_Prob'] > 1]

Empty DataFrame
Columns: [DistrictCode, SchoolCode, SchoolName, Longitude, Latitude, County, AllStudents, C_AllStudents, G_Female, GC_Female, G_Male, GC_Male, G_GenderX, GC_GenderX, R_Native, RC_Native, R_Asian, RC_Asian, R_Black, RC_Black, R_Hisp_Lat, RC_Hisp_Lat, R_HPI, RC_HPI, R_NA, RC_NA, R_TwoOrMore, RC_TwoOrMore, R_White, RC_White, L_ELL, LC_ELL, L_NoELL, LC_NoELL, I_LowIncome, IC_LowIncome, I_NoLowIncome, IC_NoLowIncome, D_Disability, DC_Disability, D_NoDisability, DC_NoDisability, A_9, AC_9, A_10, AC_10, A_11, AC_11, A_12, AC_12, PercentLowIncome, R_BIPOC, PercentBIPOC, DistrictNumSchools, DistrictSize, DistrictLowIncome, DistrictPctLowIncome, DistrictBIPOC, DistrictPctBIPOC, CS_Offered, DistrictCS, DistrictPctCS, G_Male_Prob, R_Native_Prob, R_Asian_Prob, R_Black_Prob, R_Hisp_Lat_Prob, R_HPI_Prob, R_NA_Prob, R_TwoOrMore_Prob, R_White_Prob, L_ELL_Prob, L_NoELL_Prob, I_LowIncome_Prob, I_NoLowIncome_Prob, D_Disability_Prob, D_NoDisability_Prob, A_9_Prob, A_10_Prob, A_11_Prob, A_12_Prob, GC_Female_or_X_Prob, GC_Male_Prob, RC_Native_Prob, RC_Asian_Prob, RC_Black_Prob, RC_Hisp_Lat_Prob, RC_HPI_Prob, RC_NA_Prob, RC_TwoOrMore_Prob, RC_White_Prob, LC_ELL_Prob, LC_NoELL_Prob, IC_LowIncome_Prob, IC_NoLowIncome_Prob, DC_Disability_Prob, DC_NoDisability_Prob, AC_9_Prob, AC_10_Prob, AC_11_Prob, ...]
Index: []

In [34]:
synth_students.head(50)

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   

    Latitude County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Mal

In [35]:
rng4 = np.random.default_rng(seed=3141)

def is_cs(prob):
    return rng4.choice([0, 1], p=[1 - prob, prob])

synth_students['CS_Enrolled'] = synth_students['CS_Prob'].apply(is_cs)

In [36]:
synth_students.head(50)

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   
1          1147        3015               Othello High School -119.165246   

    Latitude County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Mal

In [43]:
original_ospi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 730 entries, 0 to 729
Data columns (total 50 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   DistrictCode     730 non-null    int64  
 1   SchoolCode       730 non-null    int64  
 2   SchoolName       730 non-null    object 
 3   Longitude        730 non-null    float64
 4   Latitude         730 non-null    float64
 5   County           730 non-null    object 
 6   AllStudents      730 non-null    int64  
 7   C_AllStudents    730 non-null    int64  
 8   G_Female         730 non-null    int64  
 9   GC_Female        730 non-null    int64  
 10  G_Male           730 non-null    int64  
 11  GC_Male          730 non-null    int64  
 12  G_GenderX        730 non-null    int64  
 13  GC_GenderX       730 non-null    int64  
 14  R_Native         730 non-null    int64  
 15  RC_Native        730 non-null    int64  
 16  R_Asian          730 non-null    int64  
 17  RC_Asian        

In [42]:
sum(original_ospi.C_AllStudents) / sum(original_ospi.AllStudents)

0.06431450734579972

In data used for v1 this percentage was 7.65%...what happened?

In [44]:
sum(original_ospi.C_AllStudents)

25097

## Question: In the previous dataset the sum for this column is 29869, what happened?

There are 730 entries in both datasets. Maybe something happened to some of the C_AllStudents entries?

In [49]:
original_ospi.head()

DistrictCode  SchoolCode                        SchoolName   Longitude  \
0          1109        3075  Washtucna Elementary/High School -118.311231   
1          1147        3015               Othello High School -119.165246   
2          1158        2903        Lind-Ritzville High School -118.292516   
3          1160        2132             Ritzville High School -118.292516   
4          2250        1617    Educational Opportunity Center -117.057521   

    Latitude  County  AllStudents  C_AllStudents  G_Female  GC_Female  G_Male  \
0  46.752189   Adams           28              3         9          0      19   
1  46.822710   Adams         1281            169       635         60     646   
2  47.125588   Adams           60             17        30          8      29   
3  47.125588   Adams          113             18        57          9      56   
4  46.411019  Asotin          167              0        89          0      78   

   GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  RC_Asian  \
0        3          0           0         0          0        0         0   
1      109          0           0         0          0        5         1   
2        9          1           0         0          0        0         0   
3        9          0           0         2          2        2         0   
4        0          0           0         1          0        0         0   

   R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  RC_NA  \
0        0         0           2            0      0       0     0      0   
1        2         0        1173          151      0       0     0      0   
2        1         0          23           10      0       0     0      0   
3        0         0           9            0      0       0     0      0   
4        2         0          11            0      0       0     0      0   

   R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  L_NoELL  \
0            2             0       24         3      0       0       28   
1            3             0       98        17    402      54      879   
2            4             2       32         5      7       3       53   
3            3             0       97        16      0       0      113   
4            5             0      148         0      0       0      167   

   LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  IC_NoLowIncome  \
0         3           17             1             11               2   
1       115         1069           142            212              27   
2        14           48            14             12               3   
3        18           47             6             66              12   
4         0          133             0             34               0   

   D_Disability  DC_Disability  D_NoDisability  DC_NoDisability  A_9  AC_9  \
0             2              0              26                3    7     0   
1           195             29            1086              140  382    51   
2             3              1              57               16   18     9   
3             6              4             107               14   32     5   
4            25              0             142                0   16     0   

   A_10  AC_10  A_11  AC_11  A_12  AC_12  
0     4      0     6      1    11      2  
1   328     44   321     29   250     45  
2    15      2    14      1    13      5  
3    35      3    25      3    21      7  
4    51      0    41      0    59      0

In [52]:

v5_ospi = original_ospi.drop(columns=['SchoolName', 'Longitude', 'Latitude', 'County']).copy(deep=True)


In [55]:
v1_ospi = v1_ospi.rename(columns={'IC_NOLowIncome': 'I_NoLowIncome'})

In [54]:
v5_ospi.head()

DistrictCode  SchoolCode  AllStudents  C_AllStudents  G_Female  GC_Female  \
0          1109        3075           28              3         9          0   
1          1147        3015         1281            169       635         60   
2          1158        2903           60             17        30          8   
3          1160        2132          113             18        57          9   
4          2250        1617          167              0        89          0   

   G_Male  GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  \
0      19        3          0           0         0          0        0   
1     646      109          0           0         0          0        5   
2      29        9          1           0         0          0        0   
3      56        9          0           0         2          2        2   
4      78        0          0           0         1          0        0   

   RC_Asian  R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  \
0         0        0         0           2            0      0       0     0   
1         1        2         0        1173          151      0       0     0   
2         0        1         0          23           10      0       0     0   
3         0        0         0           9            0      0       0     0   
4         0        2         0          11            0      0       0     0   

   RC_NA  R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  \
0      0            2             0       24         3      0       0   
1      0            3             0       98        17    402      54   
2      0            4             2       32         5      7       3   
3      0            3             0       97        16      0       0   
4      0            5             0      148         0      0       0   

   L_NoELL  LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  \
0       28         3           17             1             11   
1      879       115         1069           142            212   
2       53        14           48            14             12   
3      113        18           47             6             66   
4      167         0          133             0             34   

   IC_NoLowIncome  D_Disability  DC_Disability  D_NoDisability  \
0               2             2              0              26   
1              27           195             29            1086   
2               3             3              1              57   
3              12             6              4             107   
4               0            25              0             142   

   DC_NoDisability  A_9  AC_9  A_10  AC_10  A_11  AC_11  A_12  AC_12  
0                3    7     0     4      0     6      1    11      2  
1              140  382    51   328     44   321     29   250     45  
2               16   18     9    15      2    14      1    13      5  
3               14   32     5    35      3    25      3    21      7  
4                0   16     0    51      0    41      0    59      0

In [50]:
v1_ospi.head()

DistrictCode  SchoolCode  AllStudents  C_AllStudents  G_Female  GC_Female  \
0          1109        3075           28              3         9          0   
1          1147        3015         1281            169       635         60   
2          1158        2903           60             22        30         12   
3          1160        2132          113             24        57         13   
4          2250        1617          167              0        89          0   

   G_Male  GC_Male  G_GenderX  GC_GenderX  R_Native  RC_Native  R_Asian  \
0      19        3          0           0         0          0        0   
1     646      109          0           0         0          0        5   
2      29       10          1           0         0          0        0   
3      56       11          0           0         2          2        2   
4      78        0          0           0         1          0        0   

   RC_Asian  R_Black  RC_Black  R_Hisp_Lat  RC_Hisp_Lat  R_HPI  RC_HPI  R_NA  \
0         0        0         0           2            0      0       0     0   
1         1        2         0        1173          151      0       0     0   
2         0        1         0          23           11      0       0     0   
3         0        0         0           9            0      0       0     0   
4         0        2         0          11            0      0       0     0   

   RC_NA  R_TwoOrMore  RC_TwoOrMore  R_White  RC_White  L_ELL  LC_ELL  \
0      0            2             0       24         3      0       0   
1      0            3             0       98        17    402      54   
2      0            4             2       32         9      7       3   
3      0            3             0       97        22      0       0   
4      0            5             0      148         0      0       0   

   L_NoELL  LC_NoELL  I_LowIncome  IC_LowIncome  I_NoLowIncome  \
0       28         3           17             1             11   
1      879       115         1069           142            212   
2       53        19           48            17             12   
3      113        24           47             7             66   
4      167         0          133             0             34   

   IC_NOLowIncome  D_Disability  DC_Disability  D_NoDisability  \
0               2             2              0              26   
1              27           195             29            1086   
2               5             3              1              57   
3              17             6              4             107   
4               0            25              0             142   

   DC_NoDisability  A_9  AC_9  A_10  AC_10  A_11  AC_11  A_12  AC_12  
0                3    7     0     4      0     6      1    11      2  
1              140  382    51   328     44   321     29   250     45  
2               21   18    12    15      2    14      2    13      6  
3               20   32     7    35      5    25      4    21      8  
4                0   16     0    51      0    41      0    59      0

In [47]:
v1_ospi = pd.read_csv('2022_school_pt.csv',  index_col=0)

In [57]:
print(v5_ospi.columns)
print(v1_ospi.columns)


Index(['DistrictCode', 'SchoolCode', 'AllStudents', 'C_AllStudents',
       'G_Female', 'GC_Female', 'G_Male', 'GC_Male', 'G_GenderX', 'GC_GenderX',
       'R_Native', 'RC_Native', 'R_Asian', 'RC_Asian', 'R_Black', 'RC_Black',
       'R_Hisp_Lat', 'RC_Hisp_Lat', 'R_HPI', 'RC_HPI', 'R_NA', 'RC_NA',
       'R_TwoOrMore', 'RC_TwoOrMore', 'R_White', 'RC_White', 'L_ELL', 'LC_ELL',
       'L_NoELL', 'LC_NoELL', 'I_LowIncome', 'IC_LowIncome', 'I_NoLowIncome',
       'IC_NoLowIncome', 'D_Disability', 'DC_Disability', 'D_NoDisability',
       'DC_NoDisability', 'A_9', 'AC_9', 'A_10', 'AC_10', 'A_11', 'AC_11',
       'A_12', 'AC_12'],
      dtype='object')
Index(['DistrictCode', 'SchoolCode', 'AllStudents', 'C_AllStudents',
       'G_Female', 'GC_Female', 'G_Male', 'GC_Male', 'G_GenderX', 'GC_GenderX',
       'R_Native', 'RC_Native', 'R_Asian', 'RC_Asian', 'R_Black', 'RC_Black',
       'R_Hisp_Lat', 'RC_Hisp_Lat', 'R_HPI', 'RC_HPI', 'R_NA', 'RC_NA',
       'R_TwoOrMore', 'RC_TwoOrMore', 'R_White

In [58]:
print(v5_ospi.index)
print(v1_ospi.index)


Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            720, 721, 722, 723, 724, 725, 726, 727, 728, 729],
           dtype='int64', length=730)
Int64Index([  0,   1,   2,   3,   4,   5,   6,   7,   8,   9,
            ...
            720, 721, 722, 723, 724, 725, 726, 727, 728, 729],
           dtype='int64', length=730)


In [59]:
v5_ospi_aligned, v1_ospi_aligned = v5_ospi.align(v1_ospi, join='outer', axis=1)

In [63]:
pd.set_option('display.max_rows', None)
v5_ospi_aligned[['C_AllStudents']].compare(v1_ospi_aligned[['C_AllStudents']], align_axis=1, keep_shape=False, keep_equal=False)

C_AllStudents       
             self  other
2            17.0   22.0
3            18.0   24.0
6             0.0   14.0
8           203.0  216.0
9           139.0  155.0
11            7.0   17.0
12           76.0   92.0
22           23.0  148.0
23           20.0   23.0
31            0.0    1.0
41           49.0   50.0
47            0.0   22.0
55          195.0  197.0
56          144.0  185.0
59          110.0  128.0
60           92.0  105.0
64            0.0   29.0
65            0.0   19.0
66            9.0   51.0
73           51.0   76.0
82           12.0   13.0
86            0.0   15.0
93           17.0   32.0
95           77.0  148.0
132          19.0   23.0
148          87.0  108.0
154         164.0  220.0
155         157.0  246.0
156         166.0  242.0
160         436.0  480.0
162           0.0   37.0
165          62.0   65.0
171         156.0  181.0
188          24.0   57.0
189          57.0  116.0
190          97.0  159.0
191          97.0  136.0
197         223.0  331.0
198           0.0   30.0
208         246.0  349.0
214           0.0   14.0
229           9.0   10.0
232           6.0   15.0
239           0.0    3.0
240           5.0   32.0
241           0.0   17.0
245          51.0   82.0
254          14.0   22.0
285          25.0  162.0
287           6.0  209.0
288           0.0    1.0
293          64.0  246.0
300          66.0  110.0
301         135.0  147.0
302           0.0   86.0
303           0.0  105.0
310          52.0   73.0
314          33.0   48.0
321         438.0  493.0
330         172.0  251.0
332         199.0  264.0
333          14.0   15.0
355         144.0  215.0
369          21.0   42.0
371           0.0    1.0
380          36.0   65.0
382          12.0   65.0
383          52.0   65.0
385           0.0   27.0
388           0.0   18.0
395           0.0   57.0
397          77.0  185.0
399          58.0  135.0
400          15.0   64.0
408          88.0  179.0
412           0.0  144.0
414          73.0   79.0
417           0.0   18.0
422          61.0  127.0
424          44.0  115.0
425          56.0  121.0
426          61.0  120.0
447          90.0  259.0
450         157.0  271.0
451         111.0  131.0
459         151.0  507.0
466           0.0    1.0
467         125.0  142.0
476           0.0    1.0
497          93.0  138.0
523          97.0  111.0
525         118.0  140.0
528          84.0  151.0
556          24.0   43.0
557          35.0   54.0
576           0.0    1.0
594          78.0  151.0
619           0.0    5.0
631          43.0  136.0
640          59.0  117.0
656           0.0   13.0
693           0.0  173.0
723           0.0   20.0

In [60]:
differences = v5_ospi_aligned.compare(v1_ospi_aligned, align_axis=1, keep_shape=False, keep_equal=False)
print(differences)


    AC_10       AC_11       AC_12       AC_9       C_AllStudents        \
     self other  self other  self other self other          self other   
0     NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
1     NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
2     NaN   NaN   1.0   2.0   5.0   6.0  9.0  12.0          17.0  22.0   
3     3.0   5.0   3.0   4.0   7.0   8.0  5.0   7.0          18.0  24.0   
4     NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
..    ...   ...   ...   ...   ...   ...  ...   ...           ...   ...   
725   NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
726   NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
727   NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
728   NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   
729   NaN   NaN   NaN   NaN   NaN   NaN  NaN   NaN           NaN   NaN   

    DC_Disability       DC_NoDisabili

In [56]:
v5_ospi.compare(v1_ospi, align_axis=1, keep_shape=False, keep_equal=False)

ValueError: Can only compare identically-labeled DataFrame objects

In [45]:
sum(original_ospi.AllStudents)

390223

In [46]:
original_ospi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 730 entries, 0 to 729
Data columns (total 50 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   DistrictCode     730 non-null    int64  
 1   SchoolCode       730 non-null    int64  
 2   SchoolName       730 non-null    object 
 3   Longitude        730 non-null    float64
 4   Latitude         730 non-null    float64
 5   County           730 non-null    object 
 6   AllStudents      730 non-null    int64  
 7   C_AllStudents    730 non-null    int64  
 8   G_Female         730 non-null    int64  
 9   GC_Female        730 non-null    int64  
 10  G_Male           730 non-null    int64  
 11  GC_Male          730 non-null    int64  
 12  G_GenderX        730 non-null    int64  
 13  GC_GenderX       730 non-null    int64  
 14  R_Native         730 non-null    int64  
 15  RC_Native        730 non-null    int64  
 16  R_Asian          730 non-null    int64  
 17  RC_Asian        

In [38]:
synth_students.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 322059 entries, 0 to 728
Columns: 116 entries, DistrictCode to CS_Enrolled
dtypes: float64(54), int64(54), object(8)
memory usage: 287.5+ MB


In [39]:
synth_students['CS_Enrolled'].sum()/322059

0.07676233236767176

## Preprocessing: Binning (on hold)

## Preprocessing: Normalizing
*K-Nearest Neighbors and K-means Clustering will benefit from normalizing.*

## Preprocessing: Standardizing
*Linear regression, logistic regression and principal component analysis will benefit from standardizing.*

## Preprocessing: Principle Component Analysis (PCA)

## Analysis: Random Forest

## Analysis: Logistic Regression

## Analysis: Mutual Information